Add documentation for using proxy and mirrors (#1356)
scalameta/metals: docs/build-tools/proxy.md

---
id: proxy
title: Proxy and mirrors
---

There are several parts inside Metals and Bloop that require resolving
additional dependencies from the internet. Inside environments behind a proxy or
with private artifact repositories Metals might not be able to start, because it
cannot easily download the needed dependencies.

> There a multiple ways of fixing these issues, but not all of them work in all

Reviewer: ```suggestion
There are multiple ways of fixing these issues, but not all of them work in all
```

> situations and while we are trying to make it as painless as possible to setup

Reviewer: ```suggestion
situations.  While we are trying to make it as painless as possible to setup
```

everything, there are some steps that might need to be done manually.

Everything inside Metals uses Coursier to download its dependencies:

> - Visual Studio Code extension uses Coursier bootstrap to download Metals

Reviewer: ```suggestion
- The Visual Studio Code and coc-metals extensions use Coursier bootstrap to download Metals
```

  server - Coursier boostrap is a minimal file, which only purpose is to
  download a full Coursier version. It's used to keep the extension size down to
  a minimum.
- Metals uses Coursier api to download dependencies needed for a particular
  Scala version
- Bloop uses Coursier api to download the SemanticDB plugin

> You can find some more information about particular topics on the

Reviewer: ```suggestion
You can find some more information about particular topics on
```

[Coursier's website](https://get-coursier.io/docs/overview), but we will try to

> summarize most useful solutions here.

Reviewer: ```suggestion
summarize the most useful solutions here.
```

## Mirrors

If Maven Central is not available for your workspace it might be best to use

> Coursier's mirrors. Especially that the Coursier boostrap tries to download the

Reviewer: ```suggestion
Coursier's mirrors. Especially since Coursier boostrap tries to download the
```

core of Coursier from Maven Central directly, which can only be changed using
the mirrors. Not all functionalities are yet available for the boostrapped
Coursier version, including custom repositories, so mirrors are really the only
way to fix this.

There are different ways to setup mirrors:

### Via global properties file

Create `mirror.properties` file with the contents:

```
central.from=https://repo1.maven.org/maven2
central.to=http://exmaple.com:8080/nexus/content/groups/public
```

You need to replace the uri after `central.to=` with your private repository.

The location of the mirror file varies depending on the operating system:

- Windows:
  `C:\Users\<user_name>\AppData\Roaming\Coursier\config\mirror.properties`
- Linux: `~/.config/coursier/mirror.properties`
- MacOS: `~/Library/Preferences/Coursier/mirror.properties`

This solution will work in most cases for all previously mentioned usages of

> Coursier, which is especially important in case of the Bloop server, which when

Reviewer: ```suggestion
Coursier, which is especially important in the case of the Bloop server, which when
```

started by Metals will not have any system properties forwarded.

### Via environment variable

You can set the environment variable `COURSIER_MIRRORS` with the location of the
file containing the mirror definition. That environment variable will need to

> either global or available for invocation of the Metals server. In case of the

Reviewer: ```suggestion
available either locally or globally for the Metals server. In case of the
```

Visual Studio Code extension it means you will need to start it from command
line with that variable in scope.

### Via properties

You can also use properties to specify the location of the mirror file, for
example:

```
-Dcoursier.mirrors=~/.config/coursier/mirror.properties
```

This property needs to be added to the Metals server and Coursier invocation.
However, this solution will not work for Bloop, because properties are not
forwarded to the invocation of the Bloop server. You will need to use a
different method for Bloop in this case or you can run the build server manually
with everything specified separately.

## Custom artifact repositories (Maven or Ivy resolvers)

In case you need to add custom repositories to resolve Metals server artifacts
you can use the `COURSIER_REPOSITORIES` environment variable. This will tell
Coursier to try to download artifacts from your private artifact repository.

> This is also available as a setting in the Metals Visual Studio Code extension.

Reviewer: ```suggestion
This is also available as a setting in the Metals Visual Studio Code and coc-metals extensions.
```

## Proxy settings

> In some cases, workspaces might require proxy in order to resolve the needed

Reviewer: ```suggestion
In some cases, workspaces might require a proxy in order to resolve the needed
```

> artifacts. Depending on the way Metals server is started proxy settings can be

Reviewer: ```suggestion
artifacts. Depending on the way Metals server is started, proxy settings can be
```

> specified using properties inside `.jvmopts` file and `metals.serverProperties`

Reviewer: ```suggestion
specified using properties inside a `.jvmopts` file and `metals.serverProperties`
```

> (for Visual Studio Code extension) or via properties for Coursier and Metals

Reviewer: ```suggestion
(for Visual Studio Code or coc-metals extensions) or via properties for Coursier and Metals
```

invocations.

However, because proxy properties might vary between workspaces and a Bloop
server must work for multiple clients at the same time they are not forwarded to
the Bloop itself. To make sure that Bloop uses the correct proxy settings, you
can specify them in two ways:

> - via copying the correct `.jvmopts` file to `~/.bloop` directory, which will

Reviewer: ```suggestion
- via copying the correct `.jvmopts` file to the `~/.bloop` directory, which will
```

> - via copying the correct `.jvmopts` file to the `~/.bloop` directory, which will

Reviewer 2: What about just recommending this option? When the server is started, proxy settings should just work. I would then tell users that if they have different proxy settings per workspace, every time they run `bloop` on those will reapply the proxy settings as they are global. This shouldn't be a big deal in most cases because Bloop only uses these settings to download the compiler bridges and the semanticdb, which are usually downloaded the first time a compilation happens.

Author: Thanks! I left the single way to define proxy settings and added more details about it.

  make them global
- specifying environment variable for a correct proxy type:
  - `http_proxy=<hostname>:<port>`
  - `https_proxy=<hostname>:<port>`
  - `ftp_proxy=<hostname>:<port>`
  - `socket_proxy=<hostname>:<port>`

## Troubleshooting

> There has been already a couple of issues surrounding this topic, if you are

Reviewer: ```suggestion
There has already been a couple of issues surrounding this topic, if you are
```

### The server did not start, got FailedToOpenBspConnection

```
error: The command bsp --protocol tcp --port 44559 returned with an error
>
error: The launcher failed to establish a bsp connection, aborting...
ERROR Failed to connect with build server, no functionality will work.
java.lang.RuntimeException: The server did not start, got FailedToOpenBspConnection
	at bloop.launcher.LauncherMain.failPromise$1(Launcher.scala:92)
	at bloop.launcher.LauncherMain.runLauncher(Launcher.scala:119)
	at scala.meta.internal.metals.BloopServers$$anon$1.run(BloopServers.scala:101)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.lang.Thread.run(Thread.java:834)
```

```
Exception in thread "bloop-server-background" java.io.IOException: Cannot run program "java" (in directory "<some path here ...>"): error=2, No such file or directory
        at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
```

> This might mean that you don't have your java installation on PATH and Bloop was

Reviewer: ```suggestion
This might mean that you don't have your java installation on your PATH and Bloop was
```
